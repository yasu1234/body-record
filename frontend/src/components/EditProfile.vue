<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import Cookies from 'js-cookie';

import DropFile from './DropFile.vue'
import Header from './Header.vue'

const goalWeight = ref(null);
const goalFatPercentage = ref(null);
const files = ref([]);
const userThumbnail = ref(null);

function onFileChange(event) {
    files.value = [...event];
}

const registerRecord = async () => {

}
</script>

<template>
    <Header />
    <form @submit.prevent="login">
        <div class="profile-edit-content">
            <label for="goal-weight">目標体重:</label>
            <input type="text" id="goal-weight">
            <label for="goal-weight" class="unit-label">kg</label>
        </div>
        <div class="profile-edit-content">
            <label for="goal-fat-percentage">目標体脂肪率:</label>
            <input type="text" id="goal-fat-percentage">
            <label for="goal-fat-percentage" class="unit-label">%</label>
        </div>
        <div class="profile-area">
            <label for="comments">紹介文</label>
            <textarea id="profile"></textarea>
        </div>
        <div class="profile-image-change-container">
            <label>プロフィール画像変更</label>
            <div class="current-thumbnail">
                <img v-if="userThumbnail !== null" :src="userThumbnail.url" alt="ユーザーアイコン"/>
                <img v-else src="../assets/image/user-placeholder.png" alt="ユーザーアイコン" />
            </div>
            <DropFile @change="onFileChange" />
        </div>
        <button class="edit-profile-button">更新する</button>
    </form>
</template>

<style>
form {
    margin-left: 15px;
}

.profile-edit-content {
    display: flex;
    align-items: center;
    margin-top: 20px;
}

input[type="text"] {
    padding: 5px;
    border: 1px solid #ccc;
    border-radius: 3px;
    font-size: 16px;
    width: 100px;
    margin-left: 10px;
}

.unit-label {
    margin-left: 10px;
}

textarea {
    width: 60%;
    height: 100px;
}

.profile-area {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    margin-top: 20px;
}

.profile-area label {
    margin-bottom: 5px;
}

.edit-profile-button {
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
    cursor: pointer;
    display: block;
    width: 180px;
    margin: 0 auto;
    margin-top: 30px;
}
.profile-image-change-container {
    margin-top:20px;
}

.current-thumbnail {
    width: 80px;
    height: 80px;
    border: #CCC 1px solid;
    border-radius: 55px;
    margin-top: 20px;
    margin-bottom: 20px;
}

.current-thumbnail img {
  max-width: 100%;
  max-height: 100%;
  object-fit: contain;
  border-radius: 50%;
}
</style>