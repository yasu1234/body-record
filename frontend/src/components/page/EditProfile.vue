<script setup>
import { ref, onMounted } from "vue";
import { useRoute, useRouter } from "vue-router";
import axios from "axios";
import Cookies from "js-cookie";
import FloatLabel from "primevue/floatlabel";
import InputText from "primevue/inputtext";
import Textarea from "primevue/textarea";
import Toast from "primevue/toast";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../const/toast.js";

import DropFile from "../atom/DropFile.vue";
import Header from "../layout/Header.vue";
import SettingSideMenu from "../layout/SettingSideMenu.vue";
import TabMenu from "../layout/TabMenu.vue";

const route = useRoute();
const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);

const goalWeight = ref(null);
const goalFatPercentage = ref(null);
const profile = ref('');
const file = ref(null);
const userThumbnail = ref(null);
const userId = ref(0);

function onFileChange(event) {
  file.value = event[0];
}

onMounted(() => {
  getProfile();
});

const getProfile = async () => {
  userId.value = route.params.id;
  const id = route.params.id;

  try {
    const res = await axios.get(
      import.meta.env.VITE_APP_API_BASE + `/api/v1/profiles/${id}`,
      {
        headers: {
          "access-token": Cookies.get("accessToken"),
          client: Cookies.get("client"),
          uid: Cookies.get("uid"),
        },
      }
    );

    if (res.data.user.goal_weight !== null) {
      goalWeight.value = res.data.user.goal_weight;
    }

    if (res.data.user.goal_fat_percentage !== null) {
      goalFatPercentage.value = res.data.user.goal_fat_percentage;
    }

    if (res.data.user.profile !== null) {
      profile.value = res.data.user.profile;
    }

    if (res.data.user.user.image_url) {
      userThumbnail.value = res.data.user.user.image_url;
    }
  } catch (error) {
    console.log({ error });
  }
};

const updateProfile = async () => {
  try {
    const formData = new FormData();
    formData.append("profile", profile.value);
    formData.append("goal_weight", goalWeight.value);
    formData.append("goal_fat_percentage", goalFatPercentage.value);

    const res = await axios.post(
      import.meta.env.VITE_APP_API_BASE + `/api/v1/profiles`,
      formData,
      {
        headers: {
          "access-token": Cookies.get("accessToken"),
          client: Cookies.get("client"),
          uid: Cookies.get("uid"),
        },
      }
    );
    if (file.value !== null) {
      updateProfileImage();
    } else {
      toastNotifications.displayInfo("プロフィールを更新しました", "");
      setTimeout(async () => {
        showProfile(res.data.user.id);
      }, 3000);
    }
  } catch (error) {
    let errorMessages = "";
    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      }
    }
    toastNotifications.displayError("プロフィール変更に失敗しました", errorMessages);
  }
};

const updateProfileImage = async () => {
  try {
    const formData = new FormData();
    if (file.value !== null) {
      formData.append("image", file.value);
    }

    const res = await axios.put(
      import.meta.env.VITE_APP_API_BASE + `/api/v1/auth`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
          "access-token": Cookies.get("accessToken"),
          client: Cookies.get("client"),
          uid: Cookies.get("uid"),
        },
      }
    );
    
    toastNotifications.displayInfo("プロフィールを更新しました", "");
    setTimeout(async () => {
      showProfile(res.data.user.id);
    }, 3000);
  } catch (error) {
    let errorMessages = "";
    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      }
    }
    toastNotifications.displayError("プロフィール画像登録に失敗しました", errorMessages);
  }
};

const showProfile = (id) => {
  router.push({ name: "UserProfile", params: { id: id } });
}
</script>

<template>
  <Header />
  <TabMenu />
  <Toast position="top-center" />
  <div class="setting-container">
    <SettingSideMenu :currentIndex="1" class="setting-side-menu" />
    <main class="setting-main">
      <form @submit.prevent="updateProfile">
        <div class="profile-edit-content">
          <FloatLabel>
            <InputText v-model="goalWeight" />
            <label>目標体重</label>
          </FloatLabel>
          <label for="goal-weight" class="unit-label">kg</label>
        </div>
        <div class="profile-edit-content">
          <FloatLabel>
            <InputText v-model="goalFatPercentage" />
            <label>目標体脂肪率</label>
          </FloatLabel>
          <label for="goal-fat-percentage" class="unit-label">%</label>
        </div>
        <div class="profile-area">
          <FloatLabel>
            <Textarea v-model="profile" rows="10" class="profile-text" />
            <label>紹介文</label>
          </FloatLabel>
        </div>
        <div class="profile-image-change-container">
          <label>プロフィール画像変更</label>
          <div class="current-thumbnail">
            <img
              v-if="userThumbnail !== null"
              :src="userThumbnail.url"
              alt="ユーザーアイコン"
            />
            <img
              v-else
              src="../../assets/image/user-placeholder.png"
              alt="ユーザーアイコン"
            />
          </div>
          <DropFile @change="onFileChange" />
        </div>
        <button class="edit-profile-button">更新する</button>
      </form>
    </main>
  </div>
</template>

<style scoped>
form {
  margin-left: 15px;
}

.profile-edit-content {
  display: flex;
  align-items: center;
  margin-top: 20px;
}

.unit-label {
  margin-left: 10px;
}

.profile-text {
  height: 100px;
  width: calc(100% - 40px);
}

.profile-area {
  margin-top: 20px;
}

.setting-side-menu {
  z-index: 30;
  flex: 1;
}
.setting-main {
  width: 100%;
}
.edit-profile-button {
  font-size: 16px;
  font-weight: bold;
  cursor: pointer;
  display: block;
  width: 180px;
  margin: 0 auto;
  margin-top: 30px;
}
.profile-image-change-container {
  margin-top: 20px;
}

.current-thumbnail {
  width: 80px;
  height: 80px;
  border: #ccc 1px solid;
  border-radius: 50%;
  margin-top: 20px;
  margin-bottom: 20px;
}

.current-thumbnail img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: 50%;
}
</style>
