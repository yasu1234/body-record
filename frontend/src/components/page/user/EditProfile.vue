<script setup>
import { ref, onMounted } from "vue";
import { useRoute, useRouter } from "vue-router";
import FloatLabel from "primevue/floatlabel";
import InputText from "primevue/inputtext";
import Textarea from "primevue/textarea";
import Toast from "primevue/toast";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../../js/toast.js";
import { axiosInstance, setupInterceptors } from "../../../js/axios.js";

import DropFile from "../../atom/DropFile.vue";
import Header from "../../layout/Header.vue";
import SettingSideMenu from "../../layout/SettingSideMenu.vue";
import TabMenu from "../../layout/TabMenu.vue";

const route = useRoute();
const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);
setupInterceptors(router)

const goalWeight = ref(null);
const goalFatPercentage = ref(null);
const profile = ref("");
const file = ref(null);
const userThumbnail = ref(null);
const userId = ref(0);

function onFileChange(event) {
  file.value = event[0];
}

onMounted(() => {
  getProfile();
});

const getProfile = async () => {
  userId.value = route.params.id;
  const id = route.params.id;

  try {
    const res = await axiosInstance.get(`/api/v1/profiles/${id}`);

    goalWeight.value = res.data.user.profile.goal_weight;
    goalFatPercentage.value = res.data.user.profile.goal_fat_percentage;

    if (res.data.user.profile.profile !== null) {
      profile.value = res.data.user.profile.profile;
    }

    userThumbnail.value = res.data.user.user.image_url;
  } catch (error) {
    console.log({ error });
  }
};

const updateProfile = async () => {
  try {
    const formData = new FormData();
    formData.append("profile", profile.value);
    formData.append("goal_weight", goalWeight.value);
    formData.append("goal_fat_percentage", goalFatPercentage.value);

    const res = await axiosInstance.post(`/api/v1/profiles`, formData);
    if (file.value !== null) {
      updateProfileImage();
    } else {
      toastNotifications.displayInfo("プロフィールを更新しました", "");
      setTimeout(async () => {
        showMyPage();
      }, 3000);
    }
  } catch (error) {
    let errorMessages = "";
    if (error.response != null && error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      } else {
        errorMessages = error.response.data.errors;
      }
    }
    toastNotifications.displayError(
      "プロフィール変更に失敗しました",
      errorMessages
    );
  }
};

const updateProfileImage = async () => {
  try {
    const formData = new FormData();
    if (file.value !== null) {
      formData.append("image", file.value);
    }

    const res = await axiosInstance.put(`/api/v1/auth`, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });

    toastNotifications.displayInfo("プロフィールを更新しました", "");
    setTimeout(async () => {
      showMyPage();
    }, 3000);
  } catch (error) {
    let errorMessages = "";
    if (error.response != null && error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      }
    }
    toastNotifications.displayError(
      "プロフィール画像登録に失敗しました",
      errorMessages
    );
  }
};

const showMyPage = () => {
  router.push({ name: "MyPage" });
};
</script>

<template>
  <Header />
  <TabMenu />
  <Toast position="top-center" />
  <div class="setting-container">
    <SettingSideMenu :currentIndex="1" class="setting-side-menu" />
    <main class="setting-main">
      <form @submit.prevent="updateProfile" class="ml-3.5 pb-5">
        <div class="profile-edit-content">
          <FloatLabel>
            <InputText v-model="goalWeight" class="p-2.5" />
            <label>目標体重</label>
          </FloatLabel>
          <label for="goal-weight" class="ml-2.5">kg</label>
        </div>
        <div class="profile-edit-content">
          <FloatLabel>
            <InputText v-model="goalFatPercentage" class="p-2.5" />
            <label>目標体脂肪率</label>
          </FloatLabel>
          <label for="goal-fat-percentage" class="ml-2.5">%</label>
        </div>
        <div class="mt-5">
          <FloatLabel>
            <Textarea v-model="profile" rows="10" class="profile-text p-2.5" />
            <label>紹介文</label>
          </FloatLabel>
        </div>
        <div class="mt-5">
          <label>プロフィール画像変更</label>
          <div class="current-thumbnail">
            <img
              v-if="userThumbnail !== null"
              :src="userThumbnail.url"
              alt="ユーザーアイコン"
            />
            <img
              v-else
              src="../../../assets/image/user-placeholder.png"
              alt="ユーザーアイコン"
            />
          </div>
          <DropFile @change="onFileChange" />
        </div>
        <button class="edit-profile-button">更新する</button>
      </form>
    </main>
  </div>
</template>

<style scoped>
.profile-edit-content {
  display: flex;
  align-items: center;
  margin-top: 20px;
}
.profile-text {
  height: 100px;
  width: calc(100% - 40px);
}
.setting-side-menu {
  z-index: 30;
  flex: 1;
}
.setting-main {
  width: 100%;
}
.edit-profile-button {
  font-size: 16px;
  font-weight: bold;
  cursor: pointer;
  display: block;
  width: 180px;
  margin: 0 auto;
  margin-top: 30px;
  padding: 10px 30px;
}
.current-thumbnail {
  width: 80px;
  height: 80px;
  border: #ccc 1px solid;
  border-radius: 50%;
  margin-top: 20px;
  margin-bottom: 20px;
}
.current-thumbnail img {
  width: 100%;
  height: 100%;
  border-radius: 50%;
}
</style>
