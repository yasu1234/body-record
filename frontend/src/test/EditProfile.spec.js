import { shallowMount } from "@vue/test-utils";
import EditProfile from "@/components/page/user/EditProfile.vue";

jest.mock("axios");
jest.mock("vue-router", () => {
  const router = {
    push: jest.fn(),
    pathname: "/",
    route: "",
    asPath: "",
    query: {},
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
    useRoute: jest.fn().mockReturnValue({
      params: {
        id: 1,
      },
    }),
  };
});

jest.mock("primevue/usetoast", () => ({
  useToast: jest.fn(() => ({
    add: jest.fn(),
    removeAll: jest.fn(),
  })),
}));

jest.mock("@/js/axios.js", () => ({
  axiosInstance: {
    get: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
  setupInterceptors: jest.fn(),
}));

describe("Contact", () => {
  let wrapper;
  let mockedAxiosInstance;

  beforeEach(() => {
    wrapper = shallowMount(EditProfile, {
      global: {
        stubs: ["router-link", "router-view"],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("SnapShotテスト初期表示(登録データなし)", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("SnapShotテスト初期表示(登録データあり)", () => {
    const mockProfile = {
      profile: {
        profile: "よろしくお願いします",
        goal_weight: 66,
        goal_fat_percentage: 13,
      },
    };

    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;

    // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
    mockedAxiosInstance.get.mockResolvedValue({
      data: { user: mockProfile },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
