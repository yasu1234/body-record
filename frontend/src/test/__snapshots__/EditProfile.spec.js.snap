// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact SnapShotテスト初期表示(登録データあり) 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <div
    class="setting-container"
  >
    <setting-side-menu-stub
      class="setting-side-menu"
      currentindex="1"
    />
    <main
      class="setting-main"
    >
      <form
        class="ml-3.5 pb-5"
      >
        <div
          class="profile-edit-content"
        >
          <float-label-stub />
          <label
            class="ml-2.5"
            for="goal-weight"
          >
            kg
          </label>
        </div>
        <div
          class="profile-edit-content"
        >
          <float-label-stub />
          <label
            class="ml-2.5"
            for="goal-fat-percentage"
          >
            %
          </label>
        </div>
        <div
          class="mt-5"
        >
          <float-label-stub />
          <div
            class="mt-2 text-right profile-input-width"
          >
            <p>
              残り200文字入力できます
            </p>
          </div>
        </div>
        <div
          class="mt-5"
        >
          <label>
            プロフィール画像変更
          </label>
          <div
            class="edit-profile-thumb-image"
          >
            <i
              class="pi pi-user"
              style="font-size: 40px; color: white;"
            />
          </div>
          <drop-file-stub
            class="mt-5"
          />
        </div>
        <button
          class="edit-profile-button"
        >
          更新する
        </button>
      </form>
    </main>
  </div>
  
</div>
`;

exports[`Contact SnapShotテスト初期表示(登録データなし) 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <div
    class="setting-container"
  >
    <setting-side-menu-stub
      class="setting-side-menu"
      currentindex="1"
    />
    <main
      class="setting-main"
    >
      <form
        class="ml-3.5 pb-5"
      >
        <div
          class="profile-edit-content"
        >
          <float-label-stub />
          <label
            class="ml-2.5"
            for="goal-weight"
          >
            kg
          </label>
        </div>
        <div
          class="profile-edit-content"
        >
          <float-label-stub />
          <label
            class="ml-2.5"
            for="goal-fat-percentage"
          >
            %
          </label>
        </div>
        <div
          class="mt-5"
        >
          <float-label-stub />
          <div
            class="mt-2 text-right profile-input-width"
          >
            <p>
              残り200文字入力できます
            </p>
          </div>
        </div>
        <div
          class="mt-5"
        >
          <label>
            プロフィール画像変更
          </label>
          <div
            class="edit-profile-thumb-image"
          >
            <i
              class="pi pi-user"
              style="font-size: 40px; color: white;"
            />
          </div>
          <drop-file-stub
            class="mt-5"
          />
        </div>
        <button
          class="edit-profile-button"
        >
          更新する
        </button>
      </form>
    </main>
  </div>
  
</div>
`;
